tests/pipelines/test_utils.py

```diff
@@ -81,6 +82,58 @@ def test_get_pipeline_config_returns_correctly_for_valid_pipeline_directory(
         )
 
 
+def test_get_pipeline_config_for_pipeline_directory_referencing_external_json(
```

Comment:
I verified this integration test fails if the changes in `mlflow/pipelines/utils/__init__.py` are undone, as the Jinja template (profile file) cannot be looked up through an absolute path:

```
~/mlflow add-read-jso…-profile-yml +1 ?6 ❯ pytest tests/pipelines/test_utils.py -k "test_get_pipeline_config_for_pipeline_directory_referencing_external_json"
==================================================================== test session starts =====================================================================
platform darwin -- Python 3.7.13, pytest-7.1.2, pluggy-1.0.0 -- /Users/sid.murching/mlflow/bin/python
cachedir: .pytest_cache
rootdir: /Users/sid.murching/mlflow, configfile: pytest.ini
plugins: localserver-0.5.0, cov-3.0.0
collected 12 items / 11 deselected / 1 selected                                                                                                              

tests/pipelines/test_utils.py::test_get_pipeline_config_for_pipeline_directory_referencing_external_json FAILED                                        [100%]

========================================================================== FAILURES ==========================================================================
_________________________________________ test_get_pipeline_config_for_pipeline_directory_referencing_external_json __________________________________________

pipeline_root_path = '/private/var/folders/qc/f06r2syd75xd2dq2fqg7q76w0000gp/T/pytest-of-sid.murching/pytest-7/test_get_pipeline_config_for_p0/test_pipeline'
profile = 'test'

    def get_pipeline_config(pipeline_root_path: str = None, profile: str = None) -> Dict[str, Any]:
        """
        Obtains a dictionary representation of the configuration for the specified pipeline.
    
        :param pipeline_root_path: The absolute path of the pipeline root directory on the local
                                   filesystem. If unspecified, the pipeline root directory is
                                   resolved from the current working directory, and an
        :raises MlflowException: If the specified ``pipeline_root_path`` is not a pipeline root
                                 directory or if ``pipeline_root_path`` is ``None`` and the current
                                 working directory does not correspond to a pipeline.
        :return: The configuration of the specified pipeline.
        """
        pipeline_root_path = pipeline_root_path or get_pipeline_root_path()
        _verify_is_pipeline_root_directory(pipeline_root_path=pipeline_root_path)
        try:
            if profile:
                profile_file_path = os.path.join(
                    pipeline_root_path, _PIPELINE_PROFILE_DIR, f"{profile}.yaml"
                )
                if not os.path.exists(profile_file_path):
                    raise MlflowException(
                        "Did not find the YAML configuration file for the specified profile"
                        f" '{profile}' at expected path '{profile_file_path}'.",
                        error_code=INVALID_PARAMETER_VALUE,
                    )
                return render_and_merge_yaml(
>                   pipeline_root_path, _PIPELINE_CONFIG_FILE_NAME, profile_file_path
                )

pipeline_root_path = '/private/var/folders/qc/f06r2syd75xd2dq2fqg7q76w0000gp/T/pytest-of-sid.murching/pytest-7/test_get_pipeline_config_for_p0/test_pipeline'
profile    = 'test'
profile_file_path = '/private/var/folders/qc/f06r2syd75xd2dq2fqg7q76w0000gp/T/pytest-of-sid.murching/pytest-7/test_get_pipeline_config_for_p0/test_pipeline/profiles/test.yaml'

/Users/sid.murching/mlflow/mlflow/pipelines/utils/__init__.py:62: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

root = '/private/var/folders/qc/f06r2syd75xd2dq2fqg7q76w0000gp/T/pytest-of-sid.murching/pytest-7/test_get_pipeline_config_for_p0/test_pipeline'
template_name = 'pipeline.yaml'
context_name = '/private/var/folders/qc/f06r2syd75xd2dq2fqg7q76w0000gp/T/pytest-of-sid.murching/pytest-7/test_get_pipeline_config_for_p0/test_pipeline/profiles/test.yaml'

    def render_and_merge_yaml(root, template_name, context_name):
        """
        Renders a Jinja2-templated YAML file based on a YAML context file, merge them, and return
        result as a dictionary.
    
        :param root: Root directory of the YAML files
        :param template_name: Name of the template file
        :param context_name: Name of the context file
        :return: Data in yaml file as dictionary
        """
        import jinja2
    
        template_path = os.path.join(root, template_name)
        context_path = os.path.join(root, context_name)
    
        for path in (template_path, context_path):
            if not pathlib.Path(path).is_file():
                raise MissingConfigException("Yaml file '%s' does not exist." % path)
    
        j2_env = jinja2.Environment(
            loader=jinja2.FileSystemLoader(root, encoding=ENCODING), undefined=jinja2.StrictUndefined
        )
    
        def from_json(input_var):
            import json
    
            with open(input_var, mode="r", encoding="utf-8") as f:
                return json.load(f)
    
        j2_env.filters["from_json"] = from_json
        # Compute final source of context file (e.g. my-profile.yml), applying Jinja filters
        # like from_json as needed to load context information from files, then load into a dict
>       context_source = j2_env.get_template(context_name).render({})

context_name = '/private/var/folders/qc/f06r2syd75xd2dq2fqg7q76w0000gp/T/pytest-of-sid.murching/pytest-7/test_get_pipeline_config_for_p0/test_pipeline/profiles/test.yaml'
context_path = '/private/var/folders/qc/f06r2syd75xd2dq2fqg7q76w0000gp/T/pytest-of-sid.murching/pytest-7/test_get_pipeline_config_for_p0/test_pipeline/profiles/test.yaml'
from_json  = <function render_and_merge_yaml.<locals>.from_json at 0x137032830>
j2_env     = <jinja2.environment.Environment object at 0x137040190>
jinja2     = <module 'jinja2' from '/Users/sid.murching/mlflow/lib/python3.7/site-packages/jinja2/__init__.py'>
path       = '/private/var/folders/qc/f06r2syd75xd2dq2fqg7q76w0000gp/T/pytest-of-sid.murching/pytest-7/test_get_pipeline_config_for_p0/test_pipeline/profiles/test.yaml'
root       = '/private/var/folders/qc/f06r2syd75xd2dq2fqg7q76w0000gp/T/pytest-of-sid.murching/pytest-7/test_get_pipeline_config_for_p0/test_pipeline'
template_name = 'pipeline.yaml'
template_path = '/private/var/folders/qc/f06r2syd75xd2dq2fqg7q76w0000gp/T/pytest-of-sid.murching/pytest-7/test_get_pipeline_config_for_p0/test_pipeline/pipeline.yaml'

/Users/sid.murching/mlflow/mlflow/utils/file_utils.py:232: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <jinja2.environment.Environment object at 0x137040190>
name = '/private/var/folders/qc/f06r2syd75xd2dq2fqg7q76w0000gp/T/pytest-of-sid.murching/pytest-7/test_get_pipeline_config_for_p0/test_pipeline/profiles/test.yaml'
parent = None, globals = None

    @internalcode
    def get_template(
        self,
        name: t.Union[str, "Template"],
        parent: t.Optional[str] = None,
        globals: t.Optional[t.MutableMapping[str, t.Any]] = None,
    ) -> "Template":
        """Load a template by name with :attr:`loader` and return a
        :class:`Template`. If the template does not exist a
        :exc:`TemplateNotFound` exception is raised.
    
        :param name: Name of the template to load. When loading
            templates from the filesystem, "/" is used as the path
            separator, even on Windows.
        :param parent: The name of the parent template importing this
            template. :meth:`join_path` can be used to implement name
            transformations with this.
        :param globals: Extend the environment :attr:`globals` with
            these extra variables available for all renders of this
            template. If the template has already been loaded and
            cached, its globals are updated with any new items.
    
        .. versionchanged:: 3.0
            If a template is loaded from cache, ``globals`` will update
            the template's globals instead of ignoring the new values.
    
        .. versionchanged:: 2.4
            If ``name`` is a :class:`Template` object it is returned
            unchanged.
        """
        if isinstance(name, Template):
            return name
        if parent is not None:
            name = self.join_path(name, parent)
    
>       return self._load_template(name, globals)

globals    = None
name       = '/private/var/folders/qc/f06r2syd75xd2dq2fqg7q76w0000gp/T/pytest-of-sid.murching/pytest-7/test_get_pipeline_config_for_p0/test_pipeline/profiles/test.yaml'
parent     = None
self       = <jinja2.environment.Environment object at 0x137040190>

/Users/sid.murching/mlflow/lib/python3.7/site-packages/jinja2/environment.py:1010: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <jinja2.environment.Environment object at 0x137040190>
name = '/private/var/folders/qc/f06r2syd75xd2dq2fqg7q76w0000gp/T/pytest-of-sid.murching/pytest-7/test_get_pipeline_config_for_p0/test_pipeline/profiles/test.yaml'
globals = None

    @internalcode
    def _load_template(
        self, name: str, globals: t.Optional[t.MutableMapping[str, t.Any]]
    ) -> "Template":
        if self.loader is None:
            raise TypeError("no loader for this environment specified")
        cache_key = (weakref.ref(self.loader), name)
        if self.cache is not None:
            template = self.cache.get(cache_key)
            if template is not None and (
                not self.auto_reload or template.is_up_to_date
            ):
                # template.globals is a ChainMap, modifying it will only
                # affect the template, not the environment globals.
                if globals:
                    template.globals.update(globals)
    
                return template
    
>       template = self.loader.load(self, name, self.make_globals(globals))

cache_key  = (<weakref at 0x136f38f50; to 'FileSystemLoader' at 0x137021810>, '/private/var/folders/qc/f06r2syd75xd2dq2fqg7q76w0000gp/T/pytest-of-sid.murching/pytest-7/test_get_pipeline_config_for_p0/test_pipeline/profiles/test.yaml')
globals    = None
name       = '/private/var/folders/qc/f06r2syd75xd2dq2fqg7q76w0000gp/T/pytest-of-sid.murching/pytest-7/test_get_pipeline_config_for_p0/test_pipeline/profiles/test.yaml'
self       = <jinja2.environment.Environment object at 0x137040190>
template   = None

/Users/sid.murching/mlflow/lib/python3.7/site-packages/jinja2/environment.py:969: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <jinja2.loaders.FileSystemLoader object at 0x137021810>, environment = <jinja2.environment.Environment object at 0x137040190>
name = '/private/var/folders/qc/f06r2syd75xd2dq2fqg7q76w0000gp/T/pytest-of-sid.murching/pytest-7/test_get_pipeline_config_for_p0/test_pipeline/profiles/test.yaml'
globals = ChainMap({}, {'range': <class 'range'>, 'dict': <class 'dict'>, 'lipsum': <function generate_lorem_ipsum at 0x137111f8...<class 'jinja2.utils.Cycler'>, 'joiner': <class 'jinja2.utils.Joiner'>, 'namespace': <class 'jinja2.utils.Namespace'>})

    @internalcode
    def load(
        self,
        environment: "Environment",
        name: str,
        globals: t.Optional[t.MutableMapping[str, t.Any]] = None,
    ) -> "Template":
        """Loads a template.  This method looks up the template in the cache
        or loads one by calling :meth:`get_source`.  Subclasses should not
        override this method as loaders working on collections of other
        loaders (such as :class:`PrefixLoader` or :class:`ChoiceLoader`)
        will not call this method but `get_source` directly.
        """
        code = None
        if globals is None:
            globals = {}
    
        # first we try to get the source for this template together
        # with the filename and the uptodate function.
>       source, filename, uptodate = self.get_source(environment, name)

code       = None
environment = <jinja2.environment.Environment object at 0x137040190>
globals    = ChainMap({}, {'range': <class 'range'>, 'dict': <class 'dict'>, 'lipsum': <function generate_lorem_ipsum at 0x137111f8...<class 'jinja2.utils.Cycler'>, 'joiner': <class 'jinja2.utils.Joiner'>, 'namespace': <class 'jinja2.utils.Namespace'>})
name       = '/private/var/folders/qc/f06r2syd75xd2dq2fqg7q76w0000gp/T/pytest-of-sid.murching/pytest-7/test_get_pipeline_config_for_p0/test_pipeline/profiles/test.yaml'
self       = <jinja2.loaders.FileSystemLoader object at 0x137021810>

/Users/sid.murching/mlflow/lib/python3.7/site-packages/jinja2/loaders.py:126: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <jinja2.loaders.FileSystemLoader object at 0x137021810>, environment = <jinja2.environment.Environment object at 0x137040190>
template = '/private/var/folders/qc/f06r2syd75xd2dq2fqg7q76w0000gp/T/pytest-of-sid.murching/pytest-7/test_get_pipeline_config_for_p0/test_pipeline/profiles/test.yaml'

    def get_source(
        self, environment: "Environment", template: str
    ) -> t.Tuple[str, str, t.Callable[[], bool]]:
        pieces = split_template_path(template)
        for searchpath in self.searchpath:
            # Use posixpath even on Windows to avoid "drive:" or UNC
            # segments breaking out of the search directory.
            filename = posixpath.join(searchpath, *pieces)
            f = open_if_exists(filename)
            if f is None:
                continue
            try:
                contents = f.read().decode(self.encoding)
            finally:
                f.close()
    
            mtime = os.path.getmtime(filename)
    
            def uptodate() -> bool:
                try:
                    return os.path.getmtime(filename) == mtime
                except OSError:
                    return False
    
            # Use normpath to convert Windows altsep to sep.
            return contents, os.path.normpath(filename), uptodate
>       raise TemplateNotFound(template)
E       jinja2.exceptions.TemplateNotFound: /private/var/folders/qc/f06r2syd75xd2dq2fqg7q76w0000gp/T/pytest-of-sid.murching/pytest-7/test_get_pipeline_config_for_p0/test_pipeline/profiles/test.yaml

environment = <jinja2.environment.Environment object at 0x137040190>
f          = None
filename   = '/private/var/folders/qc/f06r2syd75xd2dq2fqg7q76w0000gp/T/pytest-of-sid.murching/pytest-7/test_get_pipeline_config_for...d2dq2fqg7q76w0000gp/T/pytest-of-sid.murching/pytest-7/test_get_pipeline_config_for_p0/test_pipeline/profiles/test.yaml'
pieces     = ['private', 'var', 'folders', 'qc', 'f06r2syd75xd2dq2fqg7q76w0000gp', 'T', ...]
searchpath = '/private/var/folders/qc/f06r2syd75xd2dq2fqg7q76w0000gp/T/pytest-of-sid.murching/pytest-7/test_get_pipeline_config_for_p0/test_pipeline'
self       = <jinja2.loaders.FileSystemLoader object at 0x137021810>
template   = '/private/var/folders/qc/f06r2syd75xd2dq2fqg7q76w0000gp/T/pytest-of-sid.murching/pytest-7/test_get_pipeline_config_for_p0/test_pipeline/profiles/test.yaml'

/Users/sid.murching/mlflow/lib/python3.7/site-packages/jinja2/loaders.py:218: TemplateNotFound

The above exception was the direct cause of the following exception:

enter_pipeline_example_directory = PosixPath('/Users/sid.murching/mlflow/examples/pipelines/sklearn_regression')
tmp_path = PosixPath('/private/var/folders/qc/f06r2syd75xd2dq2fqg7q76w0000gp/T/pytest-of-sid.murching/pytest-7/test_get_pipeline_config_for_p0')

    def test_get_pipeline_config_for_pipeline_directory_referencing_external_json(
        enter_pipeline_example_directory, tmp_path
    ):
        pipeline_root_path = enter_pipeline_example_directory
        test_pipeline_root_path = tmp_path / "test_pipeline"
        shutil.copytree(pipeline_root_path, test_pipeline_root_path)
    
        test_pipeline_config = {
            "config1": 10,
            "config2": {
                "subconfig": ["A"],
            },
            "config3": "3",
            "config_from_profile": "{{PROFILE_CONFIG}}",
        }
        write_yaml(test_pipeline_root_path, "pipeline.yaml", test_pipeline_config, overwrite=True)
    
        # Write a profiles/test.yaml file that references keys in an external JSON file
        test_profile_name = "test"
        profile_contents = {
            "PROFILE_CONFIG": """{{ ("example-json.json" | from_json)["my-key"]["nested-key"] }}"""
        }
        os.mkdir(os.path.join(test_pipeline_root_path, "profiles"))
        write_yaml(
            test_pipeline_root_path,
            f"profiles/{test_profile_name}.yaml",
            profile_contents,
            overwrite=True,
        )
    
        # Write the external JSON file
        value_from_json = "my-value-from-external-json"
        example_json = {
            "my-key": {
                "nested-key": value_from_json,
            },
        }
        with open(os.path.join(test_pipeline_root_path, "example-json.json"), "w") as handle:
            handle.write(json.dumps(example_json))
        # Load pipeline config and assert correctness
        expected_config = {}
        expected_config.update(test_pipeline_config)
        expected_config.update(
            {
                "config_from_profile": value_from_json,
                "PROFILE_CONFIG": value_from_json,
            }
        )
        with chdir(test_pipeline_root_path):
>           assert get_pipeline_config(profile=test_profile_name) == expected_config

enter_pipeline_example_directory = PosixPath('/Users/sid.murching/mlflow/examples/pipelines/sklearn_regression')
example_json = {'my-key': {'nested-key': 'my-value-from-external-json'}}
expected_config = {'PROFILE_CONFIG': 'my-value-from-external-json', 'config1': 10, 'config2': {'subconfig': ['A']}, 'config3': '3', ...}
handle     = <_io.TextIOWrapper name='/private/var/folders/qc/f06r2syd75xd2dq2fqg7q76w0000gp/T/pytest-of-sid.murching/pytest-7/test_get_pipeline_config_for_p0/test_pipeline/example-json.json' mode='w' encoding='UTF-8'>
pipeline_root_path = PosixPath('/Users/sid.murching/mlflow/examples/pipelines/sklearn_regression')
profile_contents = {'PROFILE_CONFIG': '{{ ("example-json.json" | from_json)["my-key"]["nested-key"] }}'}
test_pipeline_config = {'config1': 10, 'config2': {'subconfig': ['A']}, 'config3': '3', 'config_from_profile': '{{PROFILE_CONFIG}}'}
test_pipeline_root_path = PosixPath('/private/var/folders/qc/f06r2syd75xd2dq2fqg7q76w0000gp/T/pytest-of-sid.murching/pytest-7/test_get_pipeline_config_for_p0/test_pipeline')
test_profile_name = 'test'
tmp_path   = PosixPath('/private/var/folders/qc/f06r2syd75xd2dq2fqg7q76w0000gp/T/pytest-of-sid.murching/pytest-7/test_get_pipeline_config_for_p0')
value_from_json = 'my-value-from-external-json'

/Users/sid.murching/mlflow/tests/pipelines/test_utils.py:134: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

pipeline_root_path = '/private/var/folders/qc/f06r2syd75xd2dq2fqg7q76w0000gp/T/pytest-of-sid.murching/pytest-7/test_get_pipeline_config_for_p0/test_pipeline'
profile = 'test'

    def get_pipeline_config(pipeline_root_path: str = None, profile: str = None) -> Dict[str, Any]:
        """
        Obtains a dictionary representation of the configuration for the specified pipeline.
    
        :param pipeline_root_path: The absolute path of the pipeline root directory on the local
                                   filesystem. If unspecified, the pipeline root directory is
                                   resolved from the current working directory, and an
        :raises MlflowException: If the specified ``pipeline_root_path`` is not a pipeline root
                                 directory or if ``pipeline_root_path`` is ``None`` and the current
                                 working directory does not correspond to a pipeline.
        :return: The configuration of the specified pipeline.
        """
        pipeline_root_path = pipeline_root_path or get_pipeline_root_path()
        _verify_is_pipeline_root_directory(pipeline_root_path=pipeline_root_path)
        try:
            if profile:
                profile_file_path = os.path.join(
                    pipeline_root_path, _PIPELINE_PROFILE_DIR, f"{profile}.yaml"
                )
                if not os.path.exists(profile_file_path):
                    raise MlflowException(
                        "Did not find the YAML configuration file for the specified profile"
                        f" '{profile}' at expected path '{profile_file_path}'.",
                        error_code=INVALID_PARAMETER_VALUE,
                    )
                return render_and_merge_yaml(
                    pipeline_root_path, _PIPELINE_CONFIG_FILE_NAME, profile_file_path
                )
            else:
                return read_yaml(pipeline_root_path, _PIPELINE_CONFIG_FILE_NAME)
        except MlflowException:
            raise
        except Exception as e:
            raise MlflowException(
                "Failed to read pipeline configuration. Please verify that the `pipeline.yaml`"
                " configuration file and the YAML configuration file for the selected profile are"
                " syntactically correct and that the specified profile provides all required values"
                " for template substitutions defined in `pipeline.yaml`.",
                error_code=INVALID_PARAMETER_VALUE,
>           ) from e
E           mlflow.exceptions.MlflowException: Failed to read pipeline configuration. Please verify that the `pipeline.yaml` configuration file and the YAML configuration file for the selected profile are syntactically correct and that the specified profile provides all required values for template substitutions defined in `pipeline.yaml`.

pipeline_root_path = '/private/var/folders/qc/f06r2syd75xd2dq2fqg7q76w0000gp/T/pytest-of-sid.murching/pytest-7/test_get_pipeline_config_for_p0/test_pipeline'
profile    = 'test'
profile_file_path = '/private/var/folders/qc/f06r2syd75xd2dq2fqg7q76w0000gp/T/pytest-of-sid.murching/pytest-7/test_get_pipeline_config_for_p0/test_pipeline/profiles/test.yaml'

/Users/sid.murching/mlflow/mlflow/pipelines/utils/__init__.py:75: MlflowException

---------- coverage: platform darwin, python 3.7.13-final-0 ----------
Coverage HTML written to dir htmlcov

==================================================================== slowest 10 durations ====================================================================
0.05s call     tests/pipelines/test_utils.py::test_get_pipeline_config_for_pipeline_directory_referencing_external_json
0.01s setup    tests/pipelines/test_utils.py::test_get_pipeline_config_for_pipeline_directory_referencing_external_json

(1 durations < 0.005s hidden.  Use -vv to show these durations.)
================================================================== short test summary info ===================================================================
FAILED tests/pipelines/test_utils.py::test_get_pipeline_config_for_pipeline_directory_referencing_external_json - mlflow.exceptions.MlflowException: Failed...
============================================================== 1 failed, 11 deselected in 3.96s ==============================================================

```